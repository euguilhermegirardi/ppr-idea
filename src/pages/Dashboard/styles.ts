import styled, { css } from 'styled-components';
import { shade } from 'polished';

interface FormProps {
  hasError: boolean;
}

export const TitleContainer = styled.div`
  background: #DFE4EE;
  border-radius: 5px;
  width: 15rem;
  height: 2.5rem;
  display: flex;
  justify-content: center;
  align-items: center;
  flex-direction: flex-start;

  &:hover {
      background: ${shade(0.2, '#DFE4EE')};
    }
`;

export const Title = styled.h1`
  font-size: 30px;
  color: #031795;
`;

export const Form = styled.form<FormProps>`
  margin-top: 40px;
  max-width: 700px;
  display: flex;
  input {
    flex: 1;
    height: 70px;
    padding: 0 24px;
    border: 0;
    border-radius: 5px 0 0 5px;
    color: #3a3a3a;
    background-color: #DFE4EE;
    border-right: 0;
    ${props =>
    props.hasError &&
    css`
        border-color: #c53030;
      `}
    &::placeholder {
      color: #a8a8b3;
    }
  }
  button {
    width: 90px;
    height: 70px;
    background-color: #fff;
    border-radius: 0 5px 5px 0;
    border: 0;
    color: #031795;
    font-weight: bold;
    transition: background 0.4s;
    &:hover {
      background: ${shade(0.2, '#DFE4EE')};
    }
  }
`;

export const Repositories = styled.div`
  margin-top: 80px;
  max-width: 700px;
  a {
    background: #fff;
    border-radius: 5px;
    width: 100%;
    padding: 24px;
    display: block;
    text-decoration: none;
    box-shadow: -6px 7px 14px -5px #a8a8b3;
    display: flex;
    align-items: center;
    transition: transform 0.5s;
    & + a {
      margin-top: 16px;
    }
    &:hover {
      transform: translateX(30px);
    }
    img {
      width: 64px;
      height: 64px;
      border-radius: 50%;
    }
    div {
      margin: 0 16px;
      flex: 1;
      strong {
        font-size: 20px;
        color: #031795;
      }
      p {
        font-size: 18px;
        color: #a8a8b3;
        margin-top: 4px;
      }
    }
    svg {
      margin-left: auto;
      color: #031795;
    }
  }
`;

export const Error = styled.span`
  display: block;
  color: #c53030;
  margin-top: 30px;
`;