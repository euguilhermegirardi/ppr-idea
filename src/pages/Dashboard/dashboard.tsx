import { FC, useState, useEffect, FormEvent } from 'react';
import { Link } from 'react-router-dom';
import { FiChevronRight } from 'react-icons/fi';
import { Title, Form, Repositories, Error } from './styles';
import { DashboardRepositoryProps } from './models/dashboardProps';
import api from '../../services/api';

const Dashboard_container: FC = () => {
    const [newRepository, setNewRepository] = useState('');
    const [inputError, setInputError] = useState('');
    const [repositories, setRepositories] = useState<DashboardRepositoryProps[]>(() => {
        const storageRepositories = localStorage.getItem(
            '@GitHubExplorer:repositories',
        );

        if (storageRepositories) {
            return JSON.parse(storageRepositories);
        }
        return [];
    });

    useEffect(() => {
        localStorage.setItem(
            '@GitHubExplorer:repositories',
            JSON.stringify(repositories),
        );
    }, [repositories]);

    async function handleAddRepository(event: FormEvent<HTMLFormElement>) {
        event.preventDefault();

        if (!newRepository) {
            setInputError('Você precisa digitar autor/nome do repositório');
            return;
        }

        try {
            const response = await api.get(`users/${newRepository}/repos`);

            const repository = response.data;
            setRepositories(repository);
            setNewRepository('');
            setInputError('');
        } catch (err) {
            setInputError('Hmm... algo deu errado ao buscar esse repositório');
        }
    }

    return (
        <>
            <Title>Explore repositórios no GitHub</Title>

            <Form hasError={!!inputError} onSubmit={handleAddRepository} data-testid="form">
                <input
                    value={newRepository}
                    onChange={e => setNewRepository(e.target.value)}
                    placeholder="Digite o nome do repositório aqui"
                    type="text"
                    aria-label="search-input"
                />
                <button type="submit">Pesquisar</button>
            </Form>

            {inputError && <Error>{inputError}</Error>}

            <Repositories>
                {repositories.map(repository => (
                    <Link
                        key={repository.full_name}
                        to={`/repositories/${repository.full_name}`}
                    >
                        <img
                            src={repository.owner.avatar_url}
                            alt={repository.owner.login}
                        />
                        <div>
                            <strong>{repository.full_name}</strong>
                            <p>{repository.description}</p>
                        </div>

                        <FiChevronRight size={20} />
                    </Link>
                ))}
            </Repositories>
        </>
    );
};

export default Dashboard_container;