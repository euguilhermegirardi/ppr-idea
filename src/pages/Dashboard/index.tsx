import React, { FC, useState, useEffect, FormEvent } from 'react';
import { Link } from 'react-router-dom';
import { FiChevronRight } from 'react-icons/fi';
import api from '../../services/api';

import { Title, Form, Repositories, Error } from './styles';
import logoImg from '../../assets/logo.svg';

interface Repository {
    full_name: string;
    description: string;
    owner: {
        login: string;
        avatar_url: string;
    };
}

const Dashboard: FC = () => {
    const [newRepository, setNewRepository] = useState('');
    const [inputError, setInputError] = useState('');
    const [repositories, setRepositories] = useState<Repository[]>(() => {
        const storageRepositories = localStorage.getItem(
            '@GitHubExplorer:repositories',
        );

        if (storageRepositories) {
            return JSON.parse(storageRepositories);
        }
        return [];
    });

    useEffect(() => {
        localStorage.setItem(
            '@GitHubExplorer:repositories',
            JSON.stringify(repositories),
        );
    }, [repositories]);

    async function handleAddRepository(event: FormEvent<HTMLFormElement>) {
        event.preventDefault();

        if (!newRepository) {
            setInputError('Você precisa digitar autor/nome do repositório');
            return;
        }

        try {
            const response = await api.get<Repository>(`repos/${newRepository}`);

            const repository = response.data;
            setRepositories([...repositories, repository]);
            setNewRepository('');
            setInputError('');
        } catch (err) {
            setInputError('Hmm... algo deu errado ao buscar esse repositório');
        }
    }

    return (
        <>
            <img src={logoImg} alt="GitHub Explorer" />
            <Title>Explore repositórios no GitHub</Title>

            <Form hasError={!!inputError} onSubmit={handleAddRepository}>
                <input
                    value={newRepository}
                    onChange={e => setNewRepository(e.target.value)}
                    placeholder="Digite o nome do repositório aqui"
                />
                <button type="submit">Pesquisar</button>
            </Form>

            {inputError && <Error>{inputError}</Error>}

            <Repositories>
                {repositories.map(repository => (
                    <Link
                        key={repository.full_name}
                        to={`/repositories/${repository.full_name}`}
                    >
                        <img
                            src={repository.owner.avatar_url}
                            alt={repository.owner.login}
                        />
                        <div>
                            <strong>{repository.full_name}</strong>
                            <p>{repository.description}</p>
                        </div>

                        <FiChevronRight size={20} />
                    </Link>
                ))}
            </Repositories>
        </>
    );
};

export default Dashboard;