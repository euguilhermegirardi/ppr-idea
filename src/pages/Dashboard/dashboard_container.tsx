import React, { FC, useState, useEffect, FormEvent } from 'react';
import Dashboard from './dashboard';
import { DashboardRepositoryProps } from './models/dashboardProps';
import api from '../../services/api';

const Dashboard_container: FC = () => {
    const [newRepository, setNewRepository] = useState('');
    const [inputError, setInputError] = useState('');
    const [repositories, setRepositories] = useState<DashboardRepositoryProps[]>(() => {
        const storageRepositories = localStorage.getItem(
            '@GitHubExplorer:repositories',
        );

        if (storageRepositories) {
            return JSON.parse(storageRepositories);
        }
        return [];
    });

    useEffect(() => {
        localStorage.setItem(
            '@GitHubExplorer:repositories',
            JSON.stringify(repositories),
        );
    }, [repositories]);

    async function handleAddRepository(event: FormEvent<HTMLFormElement>) {
        event.preventDefault();

        if (!newRepository) {
            setInputError('Você precisa digitar autor/nome do repositório');
            return;
        }

        try {
            const response = await api.get(`users/${newRepository}/repos`);

            const repository = response.data;
            setRepositories(repository);
            setNewRepository('');
            setInputError('');
        } catch (err) {
            setInputError('Hmm... algo deu errado ao buscar esse repositório');
        }
    }

    return (
        <Dashboard
            newRepository={newRepository}
            inputError={inputError}
            repositories={repositories}
            handleAddRepository={handleAddRepository}
            setNewRepository={setNewRepository}
        />
    );
};

export default Dashboard_container;