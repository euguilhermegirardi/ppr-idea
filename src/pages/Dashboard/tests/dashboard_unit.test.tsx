import { MemoryRouter } from 'react-router-dom';
import { fireEvent, render, RenderResult, screen } from '@testing-library/react';
import Dashboard from '../dashboard';

describe('Dashboard', () => {
    let element: RenderResult;

    beforeEach(() => {
        // eslint-disable-next-line testing-library/no-render-in-setup
        element = render(
            <MemoryRouter>
                <Dashboard />
            </MemoryRouter>
        );
    });

    it('should render the component', () => {
        // Arrange
        expect(screen).not.toBeNull();
    });

    it('should find the main title', () => {
        // Assert
        const title = screen.getByText(/Search for a GitHub repository profile/i)

        // Arrange
        expect(title).toBeInTheDocument();
    });

    it('should be able to type something in the input', async () => {
        // Assert
        const input = screen.getByLabelText('search-input') as HTMLInputElement;

        // Arrange
        fireEvent.change(input, { target: { value: 'Barbosa, Gabs' } });
        expect(input.value).toBe('Barbosa, Gabs');
    });

    it("should match snapshopt", () => {
        // Assert
        const { container } = element;

        // Arrange
        expect(container).toMatchSnapshot();
    });
});