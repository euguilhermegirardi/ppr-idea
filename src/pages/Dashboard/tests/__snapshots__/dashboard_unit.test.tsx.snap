// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Dashboard should match snapshopt 1`] = `
<div>
  <h1
    class="sc-bczRLJ lmUuWV"
  >
    Explore repositórios no GitHub
  </h1>
  <form
    class="sc-gsnTZi iipXii"
    data-testid="form"
  >
    <input
      aria-label="search-input"
      placeholder="Digite o nome do repositório aqui"
      type="text"
      value=""
    />
    <button
      type="submit"
    >
      Pesquisar
    </button>
  </form>
  <div
    class="sc-dkzDqf gxjgLD"
  />
</div>
`;
